### PR TITLE
add resource_type parameter to security endpoints

### LINKED ISSUE #8057: Add resource_type filter to security endpoints
Hello team,

We need to add a new  filter to the Wazuh API security endpoints to be able to filter resources by its `resource_type`.

## Tasks
- [x] Add the new filter to each security endpoint.
- [x] Create new integration tests for this feature.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [ ] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [ ] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [ ] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

### PR DESCRIPTION
**Pull request by CarlosRS9**
Hello team,

This PR closes #8057. A new parameter called `resource_type` has been added to the following security endpoints:
- `GET /security/policies`
- `GET /security/roles`
- `GET /security/rules`
- `GET /security/users`

In addition to that, the examples presents in the Spec.yaml have been updated to show the `resource_type` value. 

Finally, Security unit tests and API integration tests have been updated accordingly.

## Usage examples
**Getting every `default` user available:**
```
https://localhost:55000/security/users?resource_type=default
```
```
{
  "data": {
    "affected_items": [
      {
        "id": 1,
        "username": "wazuh",
        "allow_run_as": true,
        "roles": [
          1
        ],
        "resource_type": "default"
      },
      {
        "id": 2,
        "username": "wazuh-wui",
        "allow_run_as": true,
        "roles": [
          1
        ],
        "resource_type": "default"
      }
    ],
    "total_affected_items": 2,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All specified users were returned",
  "error": 0
}
```
**Trying to get the `protected` users when there is none available:**
```
https://localhost:55000/security/users?resource_type=protected
```
```
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "No user was returned",
  "error": 0
}
```

## Test results
```
============================================= test session starts ==============================================
platform linux -- Python 3.8.5, pytest-5.0.0, py-1.9.0, pluggy-0.13.1
rootdir: /home/carlos/GIT/wazuh/framework
plugins: metadata-1.11.0, tavern-1.0.0, html-3.1.1, testinfra-5.0.0
collected 804 items                                                                                            

framework/wazuh/tests/test_active_response.py ............                                               [  1%]
framework/wazuh/tests/test_agent.py .................................................................... [  9%]
...............................                                                                          [ 13%]
framework/wazuh/tests/test_cdb_list.py .....................................................             [ 20%]
framework/wazuh/tests/test_ciscat.py .................................                                   [ 24%]
framework/wazuh/tests/test_cluster.py .....ss                                                            [ 25%]
framework/wazuh/tests/test_decoder.py ...................................                                [ 29%]
framework/wazuh/tests/test_group.py ........                                                             [ 30%]
framework/wazuh/tests/test_logtest.py ......                                                             [ 31%]
framework/wazuh/tests/test_manager.py .................................                                  [ 35%]
framework/wazuh/tests/test_mitre.py .................................................................... [ 44%]
........................................................................................................ [ 56%]
........                                                                                                 [ 57%]
framework/wazuh/tests/test_rootcheck.py .................................................                [ 64%]
framework/wazuh/tests/test_rule.py ........................................................              [ 71%]
framework/wazuh/tests/test_sca.py .......                                                                [ 71%]
framework/wazuh/tests/test_security.py ................................................................. [ 79%]
.......................................................                                                  [ 86%]
framework/wazuh/tests/test_stats.py ................                                                     [ 88%]
framework/wazuh/tests/test_syscheck.py ........................                                          [ 91%]
framework/wazuh/tests/test_syscollector.py ............                                                  [ 93%]
framework/wazuh/tests/test_task.py ............................                                          [ 96%]
framework/wazuh/tests/test_vulnerability.py ..........................                                   [100%]
============================= 802 passed, 2 skipped, 15 warnings in 145.20 seconds =============================
```
```
test_security_GET_endpoints.tavern.yaml 
	 18 passed, 1 warnings
```

```
test_rbac_black_security_endpoints.tavern.yaml 
	 25 passed, 1 warnings

test_rbac_white_security_endpoints.tavern.yaml 
	 25 passed, 1 warnings

test_security_DELETE_endpoints.tavern.yaml 
	 15 passed, 1 warnings

test_security_POST_endpoints.tavern.yaml 
	 7 passed, 1 warnings

test_security_PUT_endpoints.tavern.yaml 
	 9 passed, 1 warnings

```

### CODE REVIEW

File: api/api/controllers/security_controller.py [hunk 1]
```diff
@@ -179,7 +179,7 @@ async def logout_user(request, pretty=False, wait_for_complete=False):
 
 
 async def get_users(request, user_ids: list = None, pretty=False, wait_for_complete=False,
-                    offset=0, limit=None, search=None, sort=None):
+                    offset=0, limit=None, search=None, sort=None, resource_type=None):
```
Comment (vicferpoy) [inline api/api/controllers/security_controller.py:182]: For consistency, add parameter type.

File: api/api/spec/spec.yaml [hunk 2]
```diff
@@ -5279,7 +5279,16 @@ components:
       schema:
         type: string
         format: xml_filename_path
-
+    resource_type:
+      in: query
+      name: resource_type
+      description: "Filter by process resource type"
```
Comment (vicferpoy) [inline api/api/spec/spec.yaml:5285]: Find another description. Keep in mind that a "security resource" could be a role, a user, a policy... This description is not accurate enough.

File: api/test/integration/test_security_GET_endpoints.tavern.yaml [hunk 3]
```diff
@@ -234,6 +234,80 @@ stages:
...
+  # GET /security/roles
+  - name: Try to show the roles of the system with a non-existent search
+    request:
+      <<: *all_roles_request
+      params:
+        search: notexist
+    response:
+      status_code: 200
+      json:
+        error: !anyint
```
Comment (vicferpoy) [inline api/test/integration/test_security_GET_endpoints.tavern.yaml:247]: On new tests we must use the correct `error` code.

File: api/test/integration/test_security_GET_endpoints.tavern.yaml [hunk 4]
```diff
@@ -234,6 +234,80 @@ stages:
...
+            - id: !anyint
+              name: !anystr
+              policies: !anything
+              rules: !anything
+              users: !anything
+              resource_type: user
+          failed_items: []
+          total_affected_items: !anyint
+          total_failed_items: 0
+
```
Comment (vicferpoy) [inline api/test/integration/test_security_GET_endpoints.tavern.yaml:310]: Missing a test with an invalid `resource_type` value.

File: api/test/integration/test_security_GET_endpoints.tavern.yaml [hunk 5]
```diff
@@ -428,6 +502,62 @@ stages:
...
+    request:
+      <<: *all_policies_request
+      params:
+        resource_type: protected
+    response:
+      status_code: 200
+      json:
+        error: !anyint
+        data:
+          affected_items: []
```
Comment (vicferpoy) [inline api/test/integration/test_security_GET_endpoints.tavern.yaml:536]: Consider adding a testing `protected` resource to test this.

File: framework/wazuh/core/utils.py [hunk 6]
```diff
@@ -345,6 +350,24 @@ def get_nested_fields(dikt, select_field):
...
+         The resource_type to filter by.
+
+     Returns
+     -------
+     result_list : list
+         Filtered array of dicts with only the selected (and required) fields as keys.
+     """
+    return [x for x in array if x['resource_type'] == resource_type]
+
+
```
Comment (vicferpoy) [inline framework/wazuh/core/utils.py:370]: Was this necessary? Maybe we could use the `q` param. This feels too specific. `process_array` is used in a lot of endpoints.

**Review by vicferpoy — CHANGES_REQUESTED** [review]

File: api/api/controllers/security_controller.py (hunk 1 above)
Comment (CarlosRS9) [reply api/api/controllers/security_controller.py]: Done

File: api/api/spec/spec.yaml (hunk 2 above)
Comment (CarlosRS9) [reply api/api/spec/spec.yaml]: Updated

File: api/test/integration/test_security_GET_endpoints.tavern.yaml (hunk 3 above)
Comment (CarlosRS9) [reply api/test/integration/test_security_GET_endpoints.tavern.yaml]: Done

File: api/test/integration/test_security_GET_endpoints.tavern.yaml (hunk 4 above)
Comment (CarlosRS9) [reply api/test/integration/test_security_GET_endpoints.tavern.yaml:310]: Added

File: api/test/integration/test_security_GET_endpoints.tavern.yaml (hunk 5 above)
Comment (CarlosRS9) [reply api/test/integration/test_security_GET_endpoints.tavern.yaml:545]: The reason no `protected` were added was to ensure the function does not return anything when specifying a resource type not present in the database. We are testing no `default` or `user` security resources are being returned instead.

File: framework/wazuh/core/utils.py (hunk 6 above)
Comment (CarlosRS9) [reply framework/wazuh/core/utils.py]: After further analysis I have removed this from utils. Instead, the filtering logic is now placed inside the `security` module, to apply it before the `process_array` call.

File: api/api/controllers/security_controller.py [hunk 7]
```diff
@@ -178,9 +178,10 @@ async def logout_user(request, pretty=False, wait_for_complete=False):
     return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
 
 
-async def get_users(request, user_ids: list = None, pretty=False, wait_for_complete=False,
-                    offset=0, limit=None, search=None, sort=None):
-    """Returns information from all system roles.
+async def get_users(request, user_ids: list = None, pretty: bool = False, wait_for_complete: bool = False,
+                    offset: int = 0, limit: int = None, search: str = None, sort: str = None,
+                    resource_type: str = None):
+    """Return information from all system roles.
```
Comment (vicferpoy) [inline api/api/controllers/security_controller.py:184]:
```suggestion
    """Return information from all system users.
```

**Review by vicferpoy — CHANGES_REQUESTED** [review]

**Review by vicferpoy — APPROVED** [review]
LGTM!

**Review by davidjiglesias — APPROVED** [review]

**Comment by davidjiglesias** [comment]
LGTM!

**Pull request merged**